enacts/calc.py

```diff
@@ -65,28 +74,30 @@ def water_balance(
     Examples
     --------
     """
-    # Get time_coord info
-    time_coord_size = daily_rain[time_coord].size
-    # Get all the rain-et deltas:
-    delta_rain_et = daily_rain - et
-    # Intializing sm
-    soil_moisture = xr.DataArray(
-        data=np.empty(daily_rain.shape),
-        dims=daily_rain.dims,
-        coords=daily_rain.coords,
-        name="soil_moisture",
-        attrs=dict(description="Soil Moisture", units="mm"),
+    # Initializing soil_moisture
+    soil_moisture = water_balance_step(
+        sminit,
+        daily_rain.isel({time_dim: 0}).expand_dims(dim=time_dim),
+        et,
+        taw,
     )
-    soil_moisture[{time_coord: 0}] = (
-        sminit + delta_rain_et.isel({time_coord: 0})
-    ).clip(0, taw)
-    # Looping on time_coord
-    for i in range(1, time_coord_size):
-        soil_moisture[{time_coord: i}] = (
-            soil_moisture.isel({time_coord: i - 1})
-            + delta_rain_et.isel({time_coord: i})
-        ).clip(0, taw)
-    water_balance = xr.Dataset().merge(soil_moisture)
+    if not reduce:
+        soil_moisture, _ = xr.align(soil_moisture, daily_rain[time_dim], join="outer")
+    # Looping on time_dim
+    for t in daily_rain[time_dim][1:]:
+        sm_t = water_balance_step(
+            soil_moisture.sel({time_dim: t - np.timedelta64(1, "D")}, drop=True),
+            daily_rain.sel({time_dim: t}).expand_dims(dim=time_dim),
+            et,
+            taw,
+        )
+        soil_moisture = sm_t if reduce else xr.where(
```

Comment:
You're still allocating a new `soil_moisture` array on each iteration. `where` creates a new array and fills it with values copied from its arguments.

In contrast, the assignment in the old  (red) lines 80-82 modifies the existing `soil_moisture` array.

Reply:
or make a more involved if here with:

if reduce:
    soil_moisture = sm_t
else:
    soil_moisture[t] = sm_t

?

Reply:
I would do the latter.

Reply:
done